Extension/src/LanguageServer/copilotCompletionContextProvider.ts

```diff
@@ -0,0 +1,185 @@
+        const budgetFactor = context.activeExperiments.get("CppToolsCopilotTimeBudget");
+        return (budgetFactor as number) !== undefined ? budgetFactor as number : CopilotCompletionContextProvider.defaultTimeBudgetFactor;
+    }
+
+    public static async Create() {
+        const copilotCompletionProvider = new CopilotCompletionContextProvider();
+        await copilotCompletionProvider.registerCopilotContextProvider();
+        return copilotCompletionProvider;
+    }
+
+    public removeFile(fileUri: string): void {
+        this.completionContextCache.delete(fileUri);
+    }
+
+    public async resolve(context: CompletionContext, copilotAborts: vscode.CancellationToken): Promise<CodeSnippet[]> {
+        const startTime = performance.now();
+        const out: Logger = getOutputChannelLogger();
+        const timeBudgetFactor = await this.fetchTimeBudgetFactor(context);
+        const telemetry = new CopilotContextTelemetry();
+        let codeSnippets: CodeSnippet[] | undefined;
+        let codeSnippetsKind: SnippetsKind = SnippetsKind.Unknown;
+        try {
+            this.completionContextCancellation.cancel();
+            this.completionContextCancellation = new vscode.CancellationTokenSource();
+            const docUri = context.documentContext.uri;
+            const cachedValue: CodeSnippet[] | undefined = this.completionContextCache.get(docUri.toString());
+            const snippetsPromise = this.getCompletionContextWithCancellation(docUri,
+                context.documentContext.offset, startTime, out, telemetry.fork(), this.completionContextCancellation.token);
+            [codeSnippets, codeSnippetsKind] = await this.waitForCompletionWithTimeoutAndCancellation(
+                snippetsPromise, cachedValue, context.timeBudget * timeBudgetFactor, copilotAborts);
+            if (codeSnippetsKind === SnippetsKind.Cancelled) {
+                const duration: number = performance.now() - startTime;
+                out.appendLine(`Copilot: getCompletionContext(): cancelled, elapsed time (ms) : ${duration}`);
+                telemetry.addCancelled();
+                telemetry.addCancellationElapsed(duration);
+                throw new CancellationError();
+            }
+            telemetry.addSnippetCount(codeSnippets?.length);
+            return codeSnippets ?? [];
+        } catch (e: any) {
+            telemetry.addError();
+            throw e;
+        } finally {
+            telemetry.addKind(codeSnippetsKind.toString());
+            const duration: number = performance.now() - startTime;
+            if (codeSnippets === undefined) {
+                out.appendLine(`Copilot: getCompletionContext(): no snkppets provided (${codeSnippetsKind.toString()}), elapsed time (ms): ${duration}`);
+            } else {
+                out.appendLine(`Copilot: getCompletionContext(): provided ${codeSnippets?.length} snippets (${codeSnippetsKind.toString()}), elapsed time (ms): ${duration}`);

```

Comment:
Are you able to log the id of the request here? Or is that not valid here?

Reply:
Related to this I'm seeing "Copilot: getCompletionContext(): provided 148 snippets (gotFromCacheHit)" but then I see it make a request `LSP: (received) cpptools/getCompletionContext (id: 641)` but I don't understand why it's sending a request if it used the cache.

Reply:
Is the cache just used to return results faster and it still sends a request to update the cache for later?

Reply:
As mentioned elsewhere, feel free to not log the id if it's not easily available. I'm not sure if it's worth sending it in the return message just for logging -- it depends on if there's any case where the nearest server-side message may not match the TypeScript message.